### PR TITLE
support auto generation for concat

### PR DESCRIPTION
**Pull request by enkilee**
### PR types

Others
### PR changes

Others
### Description

https://github.com/PaddlePaddle/Paddle/issues/53267 - No.8 concat
support auto generation for concat

### CODE REVIEW

File: paddle/phi/api/yaml/ops.yaml [hunk 1]
```diff
@@ -445,6 +445,16 @@
     data_type : real
   backward : complex_grad
 
+- op : concat
+  args : (Tensor[] x, Scalar(int64_t) axis = 0)
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/ops.yaml]: `Scalar(int64_t)`-->`Scalar`

File: paddle/phi/api/yaml/backward.yaml [hunk 2]
```diff
@@ -318,6 +318,33 @@
...
+    param : [x]
+  invoke : concat(grad_x_grad, axis)
+  data_transform :
+    support_trans_dtype : grad_x_grad
+  kernel:
+    func : concat_double_grad
+    data_type : grad_x_grad
+
+- backward_op : concat_grad
+  forward : concat (Tensor[] x, Scalar axis) -> Tensor(out)
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/backward.yaml]: `Scalar axis=0`

File: paddle/fluid/operators/concat_op.cc
Comment (heavyrain-lzy) [inline paddle/fluid/operators/concat_op.cc:1]: concat_op.h也删除;
paddle/fluid/operators/unity_build_rule.cmake中的`concat_op.cc`删除

File: paddle/phi/api/yaml/ops.yaml (hunk 1 above)
Comment (enkilee) [reply paddle/phi/api/yaml/ops.yaml]: 收到

File: paddle/fluid/operators/concat_op.cc
Comment (enkilee) [reply paddle/fluid/operators/concat_op.cc:1]: 收到

File: paddle/phi/api/yaml/backward.yaml (hunk 2 above)
Comment (enkilee) [reply paddle/phi/api/yaml/backward.yaml]: 收到

File: paddle/phi/api/yaml/op_compat.yaml [hunk 3]
```diff
@@ -436,12 +436,12 @@
...
   backward : concat_grad
-  inputs:
-    x: X
-  outputs:
-    out: Out
-  attrs:
-    axis: axis
+  inputs :
+    x : X
+  outputs :
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/op_compat.yaml:485]: op_compat.yaml中的添加`concat_double_grad`和`drop_empty_grad`
```
- op : concat
  backward : concat_grad, concat_double_grad
  inputs:
    x: X
  outputs:
    out: Out
  attrs:
    axis: axis
  drop_empty_grad : [x_grad]
  scalar :
    axis :
      data_type : int
      tensor_name : AxisTensor
  extra :
    attrs : [bool use_mkldnn = false, bool use_quantizer = false, str mkldnn_data_type = "float32"]
```

Comment (enkilee) [reply paddle/phi/api/yaml/op_compat.yaml:485]: 收到

File: paddle/phi/api/yaml/op_compat.yaml [hunk 4]
```diff
@@ -435,13 +435,14 @@
...
-  attrs:
-    axis: axis
+  backward : concat_grad, concat_double_grad
+  inputs :
+    x : X
+  outputs :
+    out : Out
+  attrs :
+    axis : axis
+  drop_empty_grad : [x_grad]
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/op_compat.yaml:489]: 修改`paddle/fluid/operators/generator/generate_op.py`中的`parse_drop_empty_grad`函数：修改如下：
```
def parse_drop_empty_grad(op_fluid_list: list, bw_op_dict: dict):
    for op_comp_map in op_fluid_list:
        if 'drop_empty_grad' in op_comp_map:
            bw_names = [
                bw_name.split('(')[0].strip()
                for bw_name in op_comp_map['backward'].split(',')
            ]
            new_bw_names = [bw_name for bw_name in bw_names if bw_name in bw_op_dict]
            if len(new_bw_names) != 0:
                bws_has_out_grad = False
                for bw_name in bw_names:
                    # static_ops.yaml and ops.yaml use the common op_compat.yaml
                    for out_grad in op_comp_map['drop_empty_grad']:
                        if (
                            out_grad in bw_op_dict[bw_name]['output_dict']
                        ):
                            bw_op_dict[bw_name]['output_dict'][out_grad][
                                'drop_empty_grad'
                            ] = False
                            bws_has_out_grad = True
                assert (bws_has_out_grad == True), f'''
                        {bw_names} with {op_comp_map['drop_empty_grad']} is not existed in output_dict '''

```

Comment (enkilee) [reply paddle/phi/api/yaml/op_compat.yaml:489]: 收到

File: paddle/phi/api/yaml/backward.yaml [hunk 5]
```diff
@@ -318,6 +318,33 @@
...
 
+- backward_op : concat_double_grad
+  forward : concat_grad (Tensor[] x, Tensor grad_out, Scalar axis) -> Tensor[](grad_x)
+  args : (Tensor[] grad_x_grad, Scalar axis = 0)
+  output : Tensor(grad_out_grad)
+  infer_meta :
+    func : UnchangedMultiInferMeta
+    param : [x]
+  invoke : concat(grad_x_grad, axis)
+  data_transform :
```
Comment (heavyrain-lzy) [inline paddle/phi/api/yaml/backward.yaml]: `concat_double_grad`的`data_transform`以下的配置删除

Comment (enkilee) [reply paddle/phi/api/yaml/backward.yaml]: 收到

**Comment by luotao1** [comment]
close due to 
- https://github.com/PaddlePaddle/Paddle/pull/54217#issuecomment-1569555766

**Pull request closed without merging**